### ISSUE TITLE
[bug]: empty JSON schema `{}` causes unbounded number death-loop garbled output

### ISSUE THREAD

**Issue by MarinaMiao** [vllm-project/vllm#52011]
# Empty JSON schema `{}` causes unbounded number death-loop garbled output

## Summary

When a structured-output request uses an empty JSON schema `{"json": {}}` (equivalent to "accept any JSON value"), vLLM compiles it into a grammar whose `number` branch is unbounded and cannot be terminated. The model falls into a number death-loop, emitting up to `max_tokens` of pure digits (a single giant valid JSON number), instead of a normal JSON value. `reasoning_content` (thinking phase) is unaffected — only the final `content` is garbled.

This reproduces on **GPU** as well as Ascend NPU (confirmed by a colleague on GPU vLLM; sglang does **not** reproduce). So it is a vLLM upstream issue, not backend/platform-specific.

## Environment

- vLLM version: `v0.20.3.dev47+g8d52ba3cd` (branch `hpu-release/release-v0.20.2-20260601`, commit `8d52ba3cd`)
- Model: GLM-5.1 (w8a8), with `--reasoning-parser glm45` (thinking enabled)
- Backend: structured_outputs `auto` (resolves to `guidance`/`llguidance`; also reproduces with `xgrammar`/`outlines`)
- Hardware: Ascend HPU 910A3 (also reproduced on GPU)
- Deployment: prefill-decode disaggregation (pd-sep) — but the bug is not pd-sep-specific; it is in the structured-output schema compilation path

### Request

```bash
curl -s http://<endpoint>/v1/chat/completions \
  -H "Content-Type: application/json" \
  -d '{
    "model": "auto",
    "messages": [{"role": "user", "content": "你好"}],
    "structured_outputs": {"json": {}},
    "stream": false,
    "max_tokens": 51200
  }'
```

The key trigger is `structured_outputs: {"json": {}}` — an **empty JSON schema dict**. In JSON Schema semantics, `{}` (no `type` keyword) means "accept any JSON value" (object/array/string/number/boolean/null).

### Actual (buggy) result

```json
{
  "choices": [{
    "message": {
      "role": "assistant",
      "content": "1.1478652e-0852808689391498581278959278115265275173994480839003085814898587224337225458316874850289548247142149196596559984359243185529485706...",
      "reasoning_content": "收到用户发来的\"你好\"这样简单的问候，需要礼貌地回应并引导对话..."
    },
    "finish_reason": "length"
  }],
  "usage": {"completion_tokens": 51200, ...}
}
```

- `content` is a **single giant JSON number** (`-1751849...e+2011010...` in the original badcase, 118496 chars, all valid JSON number characters — only `0-9` plus one each of `-`, `.`, `e`, `+`; zero illegal JSON-number characters).
- `reasoning_content` is **completely normal** (fluent Chinese).
- `finish_reason`: `length` (ran to `max_tokens`), `completion_tokens`: 51200 (full budget consumed).
- With `max_tokens=200`, `content` is `1.1478652e-085280868939...` (200 tokens of digits), `finish_reason=length`.

### Expected result

Either:
- Reject the empty schema with a 400 error (like vLLM already does for `{"json": ""}` empty string), or
- Normalize `{}` to `{"type": "object"}` (json_object semantics, which has a `}` terminator and produces a normal `{}`).

sglang handles this case (does not produce the death-loop), suggesting vLLM is missing an empty-schema rejection/normalization step that sglang has.

## Root cause (code-level analysis)

Three compounding defects, all in `vllm/` (upstream, not vllm-ascend):

### 1. No empty-schema rejection/normalization anywhere

- `StructuredOutputsParams.__post_init__` (`vllm/sampling_params.py`, class at line 40, `__post_init__` at line 58) only validates mutual-exclusivity and "at least one constraint". An empty dict `{}` as the `json` field passes (`self.json is not None` → count=1 → valid). No special handling.
- `_validate_structured_outputs` (`vllm/sampling_params.py`, ~line 831-975) has early-reject guards for empty-string `json=""` (lines 857-861), empty choice list, empty grammar string, `json_object=False` — but **no guard for empty dict `json={}`**.

### 2. Falsy-`{}` validation bypass (the key bug)

`backend_xgrammar.py:298` (and identically `sampling_params.py:953`, `backend_outlines.py:175`) uses truthiness instead of `is not None`:

```python
# vllm/v1/structured_output/backend_xgrammar.py:298
if so_params.json:                 # ← if {}: → False! bool({}) == False
    ...
    xgr.Grammar.from_json_schema(schema)   # line 313, NEVER executed for {}
    return
```

`bool({}) == False`, so the entire json validation block is **skipped** for `{}`. `xgr.Grammar.from_json_schema({})` is never called during validation, so the empty schema is never inspected/rejected up front. It falls through to runtime compilation.

The same falsy-`{}` bug appears in 3 places:
- `vllm/v1/structured_output/backend_xgrammar.py:298`
- `vllm/sampling_params.py:953` (auto fallback path)
- `vllm/v1/structured_output/backend_outlines.py:175`

Compare with `vllm/v1/structured_output/request.py:67`, which correctly uses `if params.json is not None:` — the same library mixes truthiness and `is not None`, so `{}` behaves inconsistently between key-construction and validation.

### 3. Empty schema compiles to "any value" grammar → unbounded number

All four backends faithfully compile `{}` per JSON Schema spec (no `type` = accept any value), and the `number` branch is `digit+` with no max-length / no forced terminator:

- xgrammar: `vllm/v1/structured_output/backend_xgrammar.py:81` `compile_json_schema("{}")`
- guidance/llguidance: `vllm/v1/structured_output/backend_guidance.py:230` `grammar_from_json_schema("{}")`
- outlines: `vllm/v1/structured_output/backend_outlines.py:73` `build_regex_from_schema("{}")`
- lm-format-enforcer: same pattern

**Empirical proof with llguidance** (`LLMatcher.try_consume_tokens`, returns count of accepted tokens):

| schema | feed `-`+30 digits | keep feeding digits | digits then `}` | digits then EOS |
|---|---|---|---|---|
| `{}` (empty) | **ok=19 accepted** | **ok=6 always accepted, is_stopped=False** | **ok=0 rejected** (number branch refuses `}` terminator) | ok=1 accepted |
| `{"type":"object"}` | ok=0 rejected (only `{` accepted) | — | — | — |
| `{"type":"string"}` | ok=0 rejected (only `"` accepted) | — | — | — |

Serialized grammar for `{}`:
```json
{"grammars":[{"name":"json_schema","json_schema":{"x-guidance":{"whitespace_flexible":true}},"lark_grammar":null}],"max_tokens":null}
```
Note: no `type` field → "any value". The number branch accepts unlimited digits and **refuses `}` termination** (ok=0). Once the model samples a digit/`-` as the first content token, the FSM enters the number state and the grammar never permits `}`/whitespace to terminate — the model can only keep emitting digits or EOS, and EOS is out-competed by digit logits within `max_tokens`.

### Thinking-phase interaction (why reasoning is normal)

`should_fill_bitmask` (`vllm/v1/structured_output/__init__.py:264-285`) returns `reasoning_ended` (False during thinking) when `enable_in_reasoning=False` (default) — so the bitmask is all-ones during thinking (grammar not enforced, reasoning normal). `should_advance` (line 287-329) returns False during thinking, so `accept_tokens` is never called and the grammar stays in its initial state. When `</think>` is emitted, `reasoning_ended=True`, grammar activates from the **initial state** — and for `{}`, the initial state accepts any-value (including digits), opening the number attractor. The thinking-switch logic itself is correct; the empty schema's initial state is what opens the attractor.

## Why only `json_empty_dict` garbles, other structured outputs are fine

| schema | grammar root | termination |
|---|---|---|
| `{"type":"object","properties":{...}}` | must start `{`, end `}` | structural terminator, `is_terminated()` at `}` |
| `{"type":"string"}` | must start `"`, end `"` | terminator |
| `{"json_object": true}` | constrained to object | normal `{}` |
| `{}` (empty) | any value, **including number** | **number is `digit+` unbounded, no forced terminator** |

Non-empty schemas have a `type` giving the grammar a structural terminator (`}`/`"`/`]`). Empty schema has no `type`, so the number branch is "can continue or end at every step" — once the model leans toward continue, it never terminates.

## Suggested fix

Two independent fixes (either suffices; both recommended):

### Fix A (treat the root cause — recommended): normalize/reject empty schema

In `StructuredOutputsParams.__post_init__` (`vllm/sampling_params.py`):

```python
def __post_init__(self):
    """Validate that some fields are mutually exclusive."""
    # Empty schema {} = accept any JSON value; the number branch is unbounded
    # and cannot terminate, causing a number death-loop. Normalize to
    # json_object (constrain to object, has } terminator). Aligns sglang.
    if isinstance(self.json, dict) and len(self.json) == 0:
        self.json_object = True
        self.json = None
    count = sum([...])
```

### Fix B (remove the falsy-`{}` validation bypass): use `is not None`

In the 3 places that currently use `if so_params.json:`:
- `vllm/v1/structured_output/backend_xgrammar.py:298`
- `vllm/sampling_params.py:953`
- `vllm/v1/structured_output/backend_outlines.py:175`

change to `if so_params.json is not None:` (matching `request.py:67`), so `{}` goes through the normal validation/compilation path instead of being silently skipped.

### Fix A verified

Applied Fix A locally on the image vLLM (`vllm/sampling_params.py`, both prefill and decode pods, restarted vLLM). Same request now returns a normal JSON object:
```json
"content": "{\n  \"name\": \"system_thought\",\n  \"content\": \"收到用户发来的\"你好\"\n}"
```
`finish_reason` is a normal stop (not `length`), `completion_tokens` well under `max_tokens`. reasoning_content unchanged (still normal).

## Additional context

- Original badcase report (Allure): case `test_structured_outputs_json_empty_dict[non_stream/stream]` — both variants garbled, ran 1606s/1291s (full `max_tokens=51200`). The test passed only because its assertion was weak ("should handle or return error" — it accepted either a response or a 400, without checking the response is valid JSON).
- This is a general vLLM guided-decoding pitfall (empty schema = any), not Ascend-specific. Confirmed on GPU by a colleague; sglang does not reproduce (sglang normalizes/rejects empty schema).
- Related vLLM structured-output area: `StructuredOutputsParams` (`vllm/sampling_params.py:40`), `StructuredOutputManager` (`vllm/v1/structured_output/__init__.py`), backends in `vllm/v1/structured_output/backend_*.py`.
